File: algorithms/rQIEA.py
#!/usr/bin/python
#
# Algorithm rQIEA from section 2.3 in Gexiang's Survey
#

import sys
import math
import numpy as np
import random

import copy
import qopt.framework

# XXX
random.seed(1)
np.random.seed(1)

def sphere(x):
    # sphere function
    return sum(np.array(x)**2)

def rotation(alpha, beta, alphabest, betabest):
    xi_b = math.atan(betabest / alphabest);
    xi_ij = math.atan(beta / alpha);
    
    if xi_b > 0 and xi_ij > 0:
        if xi_b >= xi_ij:
            return 1
        else:
            return -1
    elif xi_b > 0 and xi_ij <= 0:
        return np.sign(alphabest * alpha)
    elif xi_b <= 0 and xi_ij > 0:
        return -np.sign(alphabest * alpha)
    elif xi_b <= 0 and xi_ij <= 0:
        if xi_b >= xi_ij:
            return 1
        else:
            return -1
    elif xi_b == 0 or xi_ij == 0 or abs(xi_b - np.pi/2) < 0.001 or abs(xi_b - np.pi/2) < 0.001:
        return random.choice([-1, 1])
    else:
        print 'error in rotation'


class rQIEA(qopt.framework.EA):

    def __init__(self):
        self.popsize = 20
        self.dim = 10
        self.bounds = (-100,100) # it should be an array [(-100, 100), (-100, 100), ...)] XXX
        self.minfitness = float('inf')
        self.b = None
        self.termination = False
        self.NoFE = 0
        self.MaxNoFE = 1e4
        self.Pc = 0.05

    def initialize(self):
        self.Q = np.zeros([self.popsize, 2, self.dim])
        self.P = np.zeros([self.popsize, self.dim])
        # Initialize Q(self.t)
        for i in xrange(self.popsize):
            self.Q[i][0] = np.random.random((1, self.dim)) * 2 - 1
            self.Q[i][1] = np.sqrt(1-self.Q[i][0]**2) # XXX should be positive or negative

    def evaluation(self):  # XXX jak to sie ma do frameworka?
        fvalues = []
        for ind in self.P:
            fvalues.append(self.fitness_function(ind))
        return fvalues

    def updateQ(self):
        # Update Q(self.t) using Q-gates
        for i in xrange(self.popsize):
            for j in xrange(self.dim):
                alpha = self.Q[i][0][j]
                beta = self.Q[i][1][j]
                alphabest = self.bestq[0,j]
                betabest = self.bestq[1,j]
                k = np.pi / (100 + np.mod(self.t, 100))
                theta = k * rotation(alpha, beta, alphabest, betabest)
                G = np.matrix([ \
                        [np.cos(theta), -np.sin(theta)], \
                        [np.sin(theta),  np.cos(theta)]])
                self.Q[i][:,j] = np.dot(G, np.matrix(self.Q[i])[:,j]).transpose() # actual rotation

    def recombination(self):
        # Recombination
        for i in xrange(self.popsize):
            if random.random() < self.Pc:
                q1 = random.randint(0, self.popsize - 1)
                q2 = random.randint(0, self.popsize - 1)
                h1 = random.choice(xrange(self.dim + 1))
                h2 = random.choice(xrange(self.dim + 1))
                if h2 < h1:
                    h1, h2 = h2, h1
                temp = np.matrix(self.Q[q1], copy=True)[:,h1:h2]
                np.matrix(self.Q[q1], copy=False)[:,h1:h2] = np.matrix(self.Q[q2])[:,h1:h2] # possibly swap alphas to betas also here
                np.matrix(self.Q[q2], copy=False)[:,h1:h2] = temp

    def generation(self):
        # Observe, Construct P(self.t) -- very specific to rQIEA algorithm
        for i in xrange(self.popsize):
            for j in xrange(self.dim):
                r = random.random()
                if r <= 0.5:
                    self.P[i,j] = self.Q[i][0][j]**2
                else:
                    self.P[i,j] = self.Q[i][1][j]**2
                # mapping into range of the optimization variable
                self.P[i,j] *= self.bounds[1] - self.bounds[0]
                self.P[i,j] += self.bounds[0]

        # Evaluate P(t)
        fvalues = self.evaluation()

        # Select the best solution and store it into b(t)
        self.best = min(fvalues) # minmax XXX
        self.bestq = copy.deepcopy(self.Q[fvalues.index(self.best)])

        #  # midminfitness = float('inf')
        #  # for i in xrange(self.popsize):
        #  #     fitness = self.fitness_function(self.P[i])
        #  #     if fitness < midminfitness:
        #  #         midminfitness = fitness
        #  #         angle = np.matrix(self.Q[i], copy=True)
        #  #         midx = self.P[i]

        #  # Store the best solution in self.b
        #  if midminfitness < self.minfitness:
        #      self.minfitness = midminfitness
        #      anglemin = angle
        #      self.b = np.matrix(midx, copy=True)

        #  self.best = None # XXX

        #  # check termination
        #  self.NoFE += self.popsize
        #  if self.NoFE >= self.MaxNoFE:
        #      self.termination = True
        #      continue

        self.updateQ()

        self.recombination()
        print self.t


if __name__ == '__main__':
    rqiea = rQIEA()
    # set parameters
    rqiea.popsize = 6
    rqiea.dim = 3
    rqiea.tmax = 10000 / rqiea.popsize
    rqiea.fitness_function = sphere
    print rqiea.run()
    print rqiea.best

File: C/rqiea.cpp

/*
 * rQIEA algorithm from Gexiang's survey
 */

#include <time.h>
#include <limits>
#include <cstdlib>
#include <cassert>
#include "rqiea.h"

#define Qij (Q + i * (2 * chromlen) + (2 * j))
#define Pij (P + i * (chromlen) + (j))

void RQIEA::initialize() {
	printf("Initializing RQIEA in C++\n");

	t = 0;
	bestval = std::numeric_limits<DTYPE>::max();

	for (int i = 0; i < this->popsize; i++) {
		for (int j = 0; j < this->chromlen; j++) {
			DTYPE alpha = 2.f * (1.f * rand() / RAND_MAX) - 1;
			DTYPE beta = sqrtf(1 - alpha * alpha) * ((rand() % 2) ? -1 : 1);
			Qij[0] = alpha;
			Qij[1] = beta;
		}
	}
}

void RQIEA::observe() {
	for (int i = 0; i < this->popsize; i++) {
		for (int j = 0; j < this->chromlen; j++) {
			DTYPE r = 1.f * rand() / RAND_MAX;
			if (r <= .5f) {
				*Pij = Qij[0] * Qij[0];
			} else {
				*Pij = Qij[1] * Qij[1];
			}
			*Pij *= (this->bounds[j][1] - this->bounds[j][0]);
			*Pij += this->bounds[j][0];
		}
	}
}

void RQIEA::storebest() {
	DTYPE val = std::numeric_limits<DTYPE>::max();
	//DTYPE *b;
	int i_best;
	for (int i = 0; i < popsize; i++) {
		if (fvals[i] < val) { // XXX minmax
			val = fvals[i];
			i_best = i;
			//b = P + i * chromlen;
			//bq = 
		}
	}
	if (val < bestval) { /// XXX minmax
		bestval = val;
		memcpy(best, P + i_best * chromlen, sizeof(DTYPE) * chromlen);
		memcpy(bestq, Q + i_best * (2 * chromlen), 2 * sizeof(DTYPE) * chromlen);
	}
}

void RQIEA::evaluate() {
	for (int i = 0; i < popsize; i++) {
		fvals[i] = problem->evaluator(P + i * chromlen, chromlen);
	}
}


RQIEA::DTYPE RQIEA::LUT(RQIEA::DTYPE alpha, RQIEA::DTYPE beta, RQIEA::DTYPE alphabest, RQIEA::DTYPE betabest) {
	DTYPE eps = 1e-5;
	DTYPE xi = atan(beta / alpha);
	DTYPE xi_b = atan(betabest / alphabest);
	if (abs(xi_b) < eps || abs(xi) < eps // xi_b or xi = 0
			|| abs(xi_b - M_PI_2) < eps || abs(xi_b - M_PI_2) < eps // xi_b or xi = pi/2
			|| abs(xi_b + M_PI_2) < eps || abs(xi_b + M_PI_2) < eps) // xi_b or xi = -pi/2
	{
		return rand() % 2 ? -1 : 1;
	}

	if (xi_b > 0 && xi > 0) {
		return xi_b >= xi ? 1 : -1;
	}
	else if (xi_b > 0 && xi < 0) {
		return sign(alpha * alphabest);
	}
	else if (xi_b < 0 && xi > 0) {
		return -sign(alpha * alphabest);
	}
	else if (xi_b < 0 && xi < 0) {
		return xi_b >= xi ? 1 : -1;
	}
	else {
		assert(false);
	}
}

void RQIEA::update() {
	for (int i = 0; i < popsize; i++) {
		for (int j = 0; j < chromlen; j++) {
			DTYPE qprim[2];
			DTYPE *q = Qij;

			DTYPE k = M_PI / (100 + t % 100);
			DTYPE theta = k * LUT(q[0], q[1], bestq[j][0], bestq[j][1]);

			qprim[0] = q[0] * cos(theta) + q[1] * (-sin(theta));
			qprim[1] = q[0] * sin(theta) + q[1] * (cos(theta));

			Qij[0] = qprim[0];
			Qij[1] = qprim[1];
		}
	}
}

void RQIEA::recombine() {
	assert(popsize > 1);
	int i, j;
	i = int(1.f * popsize * rand() / RAND_MAX);
	do
		j = int(1.f * popsize * rand() / RAND_MAX);
	while (i == j);
	int h1, h2;
	h1 = int(1.f * chromlen * rand() / RAND_MAX);
	h2 = int(1.f * (chromlen - h1) * rand() / RAND_MAX) + h1;
	//printf("%d %d\n", h1, h2);
	double *q1 = Q + i * (chromlen * 2);
	double *q2 = Q + j * (chromlen * 2);

	DTYPE buf[2 * chromlen];
	memcpy(buf, q1, 2 * chromlen * sizeof(DTYPE));

	memcpy(q1 + h1 * 2, q2 + h1, (h2 - h1) * 2);
	memcpy(q2 + h1 * 2, buf + h1, (h2 - h1) * 2);

	for (int k = h1; k < h2; k++) { // < ???
		DTYPE tmp = q1[k * 2];
		q1[k * 2] = q2[k * 2];
		q2[k * 2] = tmp;
	}
}

void RQIEA::catastrophe() {
	// not implemented
}

void RQIEA::run() {
	t = 0;
	int tmax = 10000;
	initialize();
	observe(); //
	evaluate(); //
	storebest(); //
	while (t < tmax) {
		//printf("Generation %d\n", t);
		printf("bestval: %f\n", bestval);
		t++;
		observe();
		evaluate();
		storebest();
		update();
		recombine();
	}
}

#include "cec2005.h"
#include "cec2013.h"
int main() {
	srand(time(0));
	int dim = 10;
	int popsize = 10;
	RQIEA *rQIEA = new RQIEA(dim, popsize);
	for (int i = 0; i < dim; i++) {
		rQIEA->bounds[i][0] = -100;
		rQIEA->bounds[i][1] = 100;
	}

	Problem<double,double> *fun = new CEC2005(1);
	// double x[2] = {-39.3, 58.8};
	// double val = fun->evaluator(x, 2);
	// printf("-> %f\n", val);
	// return 0;
	rQIEA->problem = fun;
	rQIEA->run();
	printf("Final bestval: %f\n", rQIEA->bestval);
	printf("Final best: ");
	for (int i = 0; i < dim; i++) {
		printf("%f, ", rQIEA->best[i]);
	}
	printf("\n");
}
